src/lib/shmem/src/scmutex.rs

```diff
@@ -119,6 +121,7 @@ mod sync {
     pub struct MutPtr<T: ?Sized>(*mut T);
     #[cfg(not(loom))]
     impl<T: ?Sized> MutPtr<T> {
+        #[allow(clippy::mut_from_ref)]
         pub unsafe fn deref(&self) -> &mut T {
             unsafe { &mut *self.0 }
```

Comment:
fwiw I'm not sure this code is sound, I think it needs an UnsafeCell

Reply:
The pointer does come from a `std::cell::UnsafeCell`, but it's a little convoluted because it needs a wrapper that is compatible with loom.

For `cfg(not(loom))`:

```rust
pub struct UnsafeCell<T>(std::cell::UnsafeCell<T>);
impl<T> UnsafeCell<T> {
    // [...]

    pub fn get_mut(&self) -> MutPtr<T> {
        MutPtr(self.0.get())
    }
}

pub struct MutPtr<T: ?Sized>(*mut T);
impl<T: ?Sized> MutPtr<T> {
    pub unsafe fn deref(&self) -> &mut T {
        unsafe { &mut *self.0 }
    }

    // [...]
}
```

Reply:
I believe (but may very well be wrong) upgrading a `&*mut T` to a `&mut T` isn't sound, even if the `*mut T` came from an `std::cell::UnsafeCell`. I think MutPtr needs to wrap a `*const std::cell::UnsafeCell<T>` and call [`raw_get`](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html#method.raw_get) (and transmute) on deref

Reply:
yeah, I'm now convinced you are right

Reply:
Unresolving just so that @sporksmith can take a look.

Reply:
Ack; agreed that this should be sound and that suppressing the lint here makes sense. FWIW I believe we also exercise this path in unit tests run under miri.